### PR TITLE
Multiple subscriptions

### PR DESCRIPTION
**Pull request by roosnic1**
WPC-309. This PR moves the subscription out of the user model into its own model. User can have multiple subscription.

### CODE REVIEW

File: packages/api/src/db/subscription.ts [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+  readonly name: string
+  readonly preferredName: string
+  readonly email: string
+}
+
+export interface Subscription {
+  readonly id: string
+  readonly createdAt: Date
+  readonly modifiedAt: Date
+  readonly userID: string
```
[the remark this answers is not in the document]
Comment (roosnic1) [reply packages/api/src/db/subscription.ts:64]: yes

File: packages/editor/src/client/panel/userCheckPicker.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+export interface UserCheckPickerProps {
+  readonly list: FullUserFragment[]
+  onClose?(): void
+  onChange?(users: FullUserFragment[]): void
+}
+
+export function UserCheckPicker({list, onChange}: UserCheckPickerProps) {
+  // const {t} = useTranslation()
+
```
Comment (camilleory) [inline packages/editor/src/client/panel/userCheckPicker.tsx:14]: If this is not needed should you not remove it ?

File: packages/editor/src/client/panel/userEditPanel.tsx [hunk 3]
```diff
@@ -49,10 +48,10 @@ export function UserEditPanel({id, onClose, onSave}: UserEditPanelProps) {
   const [active, setActive] = useState(true)
   const [roles, setRoles] = useState<FullUserRoleFragment[]>([])
   const [userRoles, setUserRoles] = useState<FullUserRoleFragment[]>([])
-  const [subscription, setUserSubscription] = useState<FullUserSubscriptionFragment>()
+  // const [subscription, setUserSubscription] = useState<FullUserSubscriptionFragment>()
```
Comment (camilleory) [inline packages/editor/src/client/panel/userEditPanel.tsx:51]: Do this commented out code needs to stay there ?

Comment (penk85) [reply packages/editor/src/client/panel/userEditPanel.tsx:51]: Wondering the same :)

**Comment by elias-summermatter** [comment]
In the editor the searchbar for subscriptions does not work for me

**Comment by michael-scheurer** [comment]
The cancelUserSubscription mutation should be slightly be renamed:
- mutation should be renamed from cancelUserSubscription to cancelSubscription (consistency with new subscriptions naming)
- description of the mutation.

image

**Review by michael-scheurer — APPROVED** [review]
The cancelUserSubscription mutation should be slightly be renamed:

mutation should be renamed from cancelUserSubscription to cancelSubscription (consistency with new subscriptions naming)
description of the mutation.

**Review by michael-scheurer — CHANGES_REQUESTED** [review]
The cancelSubscription mutation throws an error.

image

This is the request content:
operationName: "Mutation"
query: "mutation Mutation($cancelUserSubscriptionId: ID!) {\n  cancelUserSubscription(id: $cancelUserSubscriptionId) {\n    ...subscription\n    __typename\n  }\n}\n\nfragment subscription on Subscription {\n  id\n  memberPlan {\n    ...memberPlan\n    __typename\n  }\n  paymentPeriodicity\n  monthlyAmount\n  autoRenew\n  startsAt\n  paidUntil\n  paymentMethod {\n    ...paymentMethod\n    __typename\n  }\n  deactivation {\n    ...deactivation\n    __typename\n  }\n  properties {\n    ...property\n    __typename\n  }\n  __typename\n}\n\nfragment memberPlan on MemberPlan {\n  id\n  name\n  slug\n  description\n  amountPerMonthMin\n  image {\n    ...image\n    __typename\n  }\n  availablePaymentMethods {\n    ...availablePaymentMethod\n    __typename\n  }\n  __typename\n}\n\nfragment availablePaymentMethod on AvailablePaymentMethod {\n  paymentMethods {\n    ...paymentMethod\n    __typename\n  }\n  paymentPeriodicities\n  forceAutoRenewal\n  __typename\n}\n\nfragment paymentMethod on PaymentMethod {\n  id\n  paymentProviderID\n  name\n  slug\n  description\n  __typename\n}\n\nfragment image on Image {\n  id\n  title\n  description\n  tags\n  source\n  author\n  license\n  fileSize\n  extension\n  mimeType\n  format\n  width\n  height\n  focalPoint {\n    ...focalPoint\n    __typename\n  }\n  url\n  transformURL\n  __typename\n}\n\nfragment focalPoint on Point {\n  x\n  y\n  __typename\n}\n\nfragment deactivation on SubscriptionDeactivation {\n  reason\n  date\n  __typename\n}\n\nfragment property on PublicProperties {\n  key\n  value\n  __typename\n}"
variables: {cancelUserSubscriptionId: "bWMv69J169XXEhEo"}


But the subscription has been cancelled in the background.

File: packages/api/src/graphql/mutation.private.ts [hunk 4]
```diff
@@ -471,13 +472,47 @@ export const GraphQLAdminMutation = new GraphQLObjectType<undefined, Context>({
...
         const {roles} = authenticate()
         authorise(CanCreateSubscription, roles)
 
         if (isTempUser(input.userID)) throw new Error('Can not update subscription with tempUser')
 
-        return dbAdapter.subscription.createSubscription({input})
+        const subscription = await dbAdapter.subscription.createSubscription({input})
+        if (!subscription) throw new Error('Subscription not created.')
+
+        // create invoice
```
Comment (roosnic1) [inline packages/api/src/graphql/mutation.private.ts:484]: This logic is already handled in `memberContext` and i recommend to leave it there. `createSubscription` should not automatically create invoices. For that we have jobs in the `memberContext`. If you need an inovice i suggest to call/create another mutation.

File: packages/api/src/db/tempUser.ts [hunk 5]
```diff
@@ -34,10 +34,14 @@ export interface TempUser {
 
 export type OptionalTempUser = TempUser | null
 
+export type UserId = string
+export type UserIdWithTempPrefix = string
+export type GenericUserId = UserId | UserIdWithTempPrefix
```
Comment (roosnic1) [inline packages/api/src/db/tempUser.ts:39]: seems a bit useless since all of them are string.

File: packages/api/src/payments/paymentProvider.ts [hunk 6]
```diff
@@ -129,31 +137,97 @@ export abstract class BasePaymentProvider implements PaymentProvider {
 
     if (!updatedPayment) throw new Error('Error while updating Payment')
 
+    // get invoice and subscription joins out of the payment
```
Comment (roosnic1) [inline packages/api/src/payments/paymentProvider.ts:140]: the whole logic with activating the user is handled in https://github.com/wepublish/wepublish/blob/f/WPC-309-multiple-subscriptions/packages/api/src/events.ts#L176

File: packages/api/src/graphql/subscription.ts [hunk 7]
```diff
@@ -50,7 +50,7 @@ export const GraphQLSubscription = new GraphQLObjectType<Subscription, Context>(
     createdAt: {type: GraphQLNonNull(GraphQLDateTime)},
     modifiedAt: {type: GraphQLNonNull(GraphQLDateTime)},
     user: {
-      type: GraphQLNonNull(GraphQLUser),
+      type: GraphQLUser,
```
Comment (roosnic1) [inline packages/api/src/graphql/subscription.ts:53]: in which case does a subscription not have an user?

**Review by roosnic1 — COMMENTED** [review]
This PR is not yet read for merging. Some of the changes made duplicate logic and hasn't been properly tested

File: packages/api/src/db/tempUser.ts (hunk 5 above)
Comment (michael-scheurer) [reply packages/api/src/db/tempUser.ts:39]: Was requested by Tomasz.

File: packages/api/src/graphql/subscription.ts (hunk 7 above)
Comment (michael-scheurer) [reply packages/api/src/graphql/subscription.ts:53]: If user gets deleted (possible in the editor).

File: packages/api/src/graphql/mutation.private.ts (hunk 4 above)
Comment (michael-scheurer) [reply packages/api/src/graphql/mutation.private.ts:484]: Will have a look at it.

File: packages/api/src/payments/paymentProvider.ts (hunk 6 above)
Comment (michael-scheurer) [reply packages/api/src/payments/paymentProvider.ts:140]: Will have a look at it.

Comment (michael-scheurer) [reply packages/api/src/payments/paymentProvider.ts:140]: Why I introduced this code: If an user calls the `registerAndReceivePayment()` mutation and then the api receives the webhook eg. from Payrexx the following line would always evaluate false cause paidAt is obviously null: https://github.com/wepublish/wepublish/blob/eea4da1b7164318659255cbda138fa1b88cdd0c0/packages/api/src/events.ts#L178

How I tested: I called the registerAndReceivePayment() mutation and then fired the webhook from Payrexx to the api. The invoice remained unpaid.

How would you solve this problem though?

File: packages/api/src/graphql/mutation.private.ts (hunk 4 above)
Comment (michael-scheurer) [reply packages/api/src/graphql/mutation.private.ts:484]: now I saw your intention and changed that.

**Review by elias-summermatter — DISMISSED** [review]
Intensively tested and fixed yesterday together with Michael. Looked fine.

**Review by michael-scheurer — APPROVED** [review]

**Pull request merged**